tensorflow_datasets/core/features/video_feature.py

```diff
@@ -197,26 +197,76 @@ def to_json_content(self) -> Json:
         'ffmpeg_extra_args': self._extra_ffmpeg_args
     }
 
+  def _generate_video_html(self, images, framerate=24, encoding_format='mp4') -> str:
+    """Converts sequence of images into video string."""
+    # CODE REDUNDANT: Could somehow use `_ffmpeg_decode`?
+    imgs = len(str(len(images)))+1  # Find number of digits in len to give names.
+    video_dir = tempfile.mkdtemp()
+    for i, img in enumerate(images):
+      f = os.path.join(video_dir, f'img{i:0{imgs}d}.png')
+      img.save(f, format='png')
+
+    ffmpeg_args = [self._ffmpeg_path, '-framerate', str(framerate), '-i',
+                   os.fspath(os.path.join(video_dir, f'img%0{imgs}d.png'))]
+    ffmpeg_stdin = None
+
+    output_pattern = os.path.join(video_dir, 'output.' + encoding_format)
+    ffmpeg_args += self._extra_ffmpeg_args
+    ffmpeg_args.append(output_pattern)
+    try:
+      # try using ffmpeg to convert pngs to mp4
+      process = subprocess.Popen(ffmpeg_args,
+                                 stdin=subprocess.PIPE,
+                                 stdout=subprocess.PIPE,
+                                 stderr=subprocess.PIPE)
+      stdout_data, stderr_data = process.communicate(ffmpeg_stdin)
+      ffmpeg_ret_code = process.returncode
+      if ffmpeg_ret_code:
+        raise ValueError(
+            'ffmpeg returned error code {}, command={}\n'
+            'stdout={}\nstderr={}\n'.format(ffmpeg_ret_code,
+                                            ' '.join(ffmpeg_args),
+                                            stdout_data,
+                                            stderr_data))
+      video = None
+      video_file = os.path.join(video_dir, 'output.mp4')
+      with tf.io.gfile.GFile(video_file, 'rb') as vf:
+        video = vf.read()
+
+      def write_buff(buff):
+        buff.write(video)
+      video_str = utils.get_base64(write_buff)
+      return (f'<video height="128" width="128" controls loop plays-inline>'
+              f'<source src="data:image/gif;base64,{video_str}"'
+              f' type="video/mp4" alt="Video"></video>')
+
+    except OSError:
+      # if ffmpeg is not installed, generate GIF using pngs
+      duration = (41*24)/framerate
```

Comment:
Shouldn' the framerate be the same for both gif and mp4 ?

Reply:
I made this relation to find duration from given framerate (according to the mentioned comment i.e, 41ms ~= 24 imgs/s). As `PIL.Image` uses duration for creating `GIF`.